packages/node-core/src/indexer/multiChainRewind.service.ts

```diff
@@ -0,0 +1,272 @@
+
+      const rewindTriggers = await getTriggers(this.sequelize, this.rewindTriggerName);
+      if (rewindTriggers.length === 0) {
+        await this.sequelize.query(`${createRewindTrigger(this.dbSchema)}`);
+      }
+
+      // Register a listener and create a schema notification sending function.
+      await this.registerPgListener();
+
+      if (this.waitRewindHeader) {
+        const rewindHeader = {...this.waitRewindHeader};
+        await reindex(rewindHeader);
+        return rewindHeader;
+      }
+    }
+  }
+
+  private async registerPgListener() {
+    if (this.pgListener) return;
+
+    // Creating a new pgClient is to avoid using the same database connection as the block scheduler,
+    // which may prevent real-time listening to rollback events.
+    this.pgListener = (await this.sequelize.connectionManager.getConnection({
+      type: 'read',
+    })) as PoolClient;
+
+    this.pgListener.on('notification', (msg) => {
+      Promise.resolve().then(async () => {
+        const eventType = msg.payload;
+        logger.info(`Received rewind event: ${eventType}, chainId: ${this.chainId}`);
+        switch (eventType) {
+          case MultiChainRewindEvent.Rewind:
+          case MultiChainRewindEvent.RewindTimestampDecreased: {
+            const {rewindTimestamp} = await this.globalModel.getGlobalRewindStatus();
+            this.waitRewindHeader = await this.searchWaitRewindHeader(rewindTimestamp);
+            this.status = RewindStatus.Rewinding;
+
+            // Trigger the rewind event, and let the fetchService listen to the message and handle the queueFlush.
+            this.eventEmitter.emit(eventType, {
+              height: this.waitRewindHeader.blockHeight,
+            } satisfies MultiChainRewindPayload);
+            break;
+          }
+          case MultiChainRewindEvent.RewindComplete:
+            // recover indexing status
+            this.waitRewindHeader = undefined;
+            this.status = RewindStatus.Normal;
+            break;
+          default:
+            throw new Error(`Unknown rewind event: ${eventType}`);
+        }
+        logger.info(`Handle success rewind event: ${eventType}, chainId: ${this.chainId}`);
+      });
+    });
+
+    await this.pgListener.query(`LISTEN "${this.rewindTriggerName}"`);
+    logger.info(`Register rewind listener success, chainId: ${this.chainId}`);
+
+    // Check whether the current state is in rollback.
+    // If a global lock situation occurs, prioritize setting it to the WaitOtherChain state. If a rollback is still required, then set it to the rewinding state.
+    const {rewindLock, rewindTimestamp} = await this.globalModel.getGlobalRewindStatus();
+    if (rewindLock) {
+      this.status = RewindStatus.WaitOtherChain;
+    }
+    if (rewindTimestamp) {
+      this.status = RewindStatus.Rewinding;
+      this.waitRewindHeader = await this.searchWaitRewindHeader(rewindTimestamp);
+    }
```

Comment:
This is implicit behaviour of this function. It's probably best to move this to its own function